vscode/src/completions/providers/fetch-and-process-completions.ts

```diff
@@ -34,54 +34,71 @@ export async function* fetchAndProcessDynamicMultilineCompletions(
         providerOptions,
         providerSpecificPostProcess,
     } = params
-    const { hotStreak, docContext, multiline } = providerOptions
+    const { hotStreak, docContext, multiline, firstCompletionTimeout } = providerOptions
 
     let hotStreakExtractor: undefined | HotStreakExtractor
 
     interface StopParams {
         completedCompletion: InlineCompletionItemWithAnalytics
         rawCompletion: string
-        isFinal: boolean
+        isFullResponse: boolean
     }
 
     function* stopStreamingAndUsePartialResponse(
         stopParams: StopParams
     ): Generator<FetchCompletionResult> {
-        const { completedCompletion, rawCompletion, isFinal } = stopParams
+        const { completedCompletion, rawCompletion, isFullResponse } = stopParams
+        addAutocompleteDebugEvent('stopStreamingAndUsePartialResponse', {
+            isFullResponse,
+            text: rawCompletion,
+        })
 
         yield {
             docContext,
             completion: {
                 ...completedCompletion,
-                stopReason: isFinal ? completedCompletion.stopReason : 'streaming-truncation',
+                stopReason: isFullResponse ? completedCompletion.stopReason : 'streaming-truncation',
             },
         }
 
+        // TODO(valery): disable hot-streak for long multline completions?
         if (hotStreak) {
             hotStreakExtractor = createHotStreakExtractor({
                 completedCompletion,
                 ...params,
             })
 
-            yield* hotStreakExtractor.extract(rawCompletion, isFinal)
+            yield* hotStreakExtractor.extract(rawCompletion, isFullResponse)
         } else {
             abortController.abort()
         }
     }
 
+    const generatorStartTime = performance.now()
+
     for await (const { completion, stopReason } of completionResponseGenerator) {
-        const isFinal = stopReason !== STOP_REASON_STREAMING_CHUNK
-        const extractCompletion = isFinal ? parseAndTruncateCompletion : canUsePartialCompletion
+        const isFirstCompletionTimeoutElapsed =
+            performance.now() - generatorStartTime >= firstCompletionTimeout
+        const isFullResponse = stopReason !== STOP_REASON_STREAMING_CHUNK
+        const shouldFirstCompletion = isFullResponse || isFirstCompletionTimeoutElapsed
```

Comment:
```suggestion
        const shouldYieldFirstCompletion = isFullResponse || isFirstCompletionTimeoutElapsed
```

Feels like there's a verb missing? 🤔